File: 1160_FindWordsThatCanBeFormedByCharacters.cpp
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
  int countCharacters(vector<string> &words, string chars) {
    int sum = 0;

    unordered_map<char, int> lookup;
    for (int i = 0; i < chars.length(); i++) {
      lookup[chars[i]]++;
    }

    unordered_map<char, int> wordLookup;
    bool valid;
    for (int i = 0; i < words.size(); i++) {
      valid = true;

      for (int j = 0; j < words[i].length(); j++) {
        wordLookup[words[i][j]]++;

        if (lookup[words[i][j]] < wordLookup[words[i][j]]) {
          valid = false;
          break;
        }
      }

      if (valid) {
        sum += words[i].length();
      }

      wordLookup.clear();
    }

    return sum;
  }
};

File: 1160_FindWordsThatCanBeFormedByCharacters.rs
use std::collections::HashMap;

// struct Solution;

impl Solution {
    pub fn count_characters(words: Vec<String>, chars: String) -> i32 {
        let mut sum = 0;

        let mut lookup = HashMap::new();
        for c in chars.chars() {
            lookup.insert(c, lookup.get(&c).unwrap_or(&0) + 1);
        }

        let mut word_lookup = HashMap::new();
        let mut valid: bool;
        for words in words.iter() {
            valid = true;

            for c in words.chars() {
                word_lookup.insert(c, word_lookup.get(&c).unwrap_or(&0) + 1);

                if lookup.get(&c).unwrap_or(&0) < word_lookup.get(&c).unwrap() {
                    valid = false;
                    break;
                }
            }

            if valid {
                sum += words.len() as i32;
            }

            word_lookup.clear();
        }

        sum
    }
}
